File: AngularJS/src/org/angular2/entities/ivy/Angular2IvyDirectiveAttribute.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.entities.ivy;

import com.intellij.lang.javascript.psi.JSType;
import com.intellij.lang.javascript.psi.ecma6.JSTypeDeclaration;
import com.intellij.model.Pointer;
import com.intellij.psi.PsiElement;
import org.angular2.entities.Angular2DirectiveAttribute;
import org.angular2.entities.Angular2EntityUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

import static com.intellij.refactoring.suggested.UtilsKt.createSmartPointer;

public class Angular2IvyDirectiveAttribute implements Angular2DirectiveAttribute {

  private final String myName;
  private final JSTypeDeclaration mySource;

  Angular2IvyDirectiveAttribute(@NotNull String name,
                                @NotNull JSTypeDeclaration source) {
    myName = name;
    mySource = source;
  }

  @Override
  public @NotNull String getName() {
    return myName;
  }

  @Override
  public @Nullable JSType getJsType() {
    return mySource.getJSType();
  }

  @Override
  public @NotNull PsiElement getSourceElement() {
    return mySource;
  }

  @Override
  public String toString() {
    return Angular2EntityUtils.toString(this);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Angular2IvyDirectiveAttribute attribute = (Angular2IvyDirectiveAttribute)o;
    return myName.equals(attribute.myName) && mySource.equals(attribute.mySource);
  }

  @Override
  public int hashCode() {
    return Objects.hash(myName, mySource);
  }

  @NotNull
  @Override
  public Pointer<Angular2IvyDirectiveAttribute> createPointer() {
    var name = myName;
    var source = createSmartPointer(mySource);
    return () -> {
      var newSource = source.getElement();
      return newSource != null ? new Angular2IvyDirectiveAttribute(name, newSource) : null;
    };
  }
}

File: Angular/src/org/angular2/entities/ivy/Angular2IvyDirectiveAttribute.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.angular2.entities.ivy

import com.intellij.javascript.webSymbols.apiStatus
import com.intellij.lang.javascript.psi.JSElementBase
import com.intellij.lang.javascript.psi.JSType
import com.intellij.lang.javascript.psi.ecma6.JSTypeDeclaration
import com.intellij.model.Pointer
import com.intellij.psi.PsiElement
import com.intellij.refactoring.suggested.createSmartPointer
import com.intellij.webSymbols.WebSymbolApiStatus
import org.angular2.entities.Angular2DirectiveAttribute
import org.angular2.entities.Angular2EntityUtils
import java.util.*

class Angular2IvyDirectiveAttribute internal constructor(override val name: String,
                                                         private val mySource: JSTypeDeclaration) : Angular2DirectiveAttribute {

  override val type: JSType
    get() = mySource.jsType

  override val sourceElement: PsiElement
    get() = mySource

  override val apiStatus: WebSymbolApiStatus
    get() = (mySource as? JSElementBase)?.apiStatus ?: WebSymbolApiStatus.Stable

  override fun toString(): String {
    return Angular2EntityUtils.toString(this)
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val attribute = other as Angular2IvyDirectiveAttribute?
    return name == attribute!!.name && mySource == attribute.mySource
  }

  override fun hashCode(): Int {
    return Objects.hash(name, mySource)
  }

  override fun createPointer(): Pointer<Angular2IvyDirectiveAttribute> {
    val name = this.name
    val source = mySource.createSmartPointer()
    return Pointer {
      source.element?.let { Angular2IvyDirectiveAttribute(name, it) }
    }
  }
}
